apps/web/app/blog/posts/ai-powered-spreadsheet/page.mdx

```diff
@@ -0,0 +1,217 @@
+  tool: updateCell,
+  toolSchema: z.function().args(
+    z.union([z.string(), z.number()]).describe("Row identifier"),
+    z.string().describe("Column identifier (e.g., 'A', 'B')"),
+    z.object({
+      type: z.literal("text"),
+      text: z.string()
+    }).or(z.object({
+      type: z.literal("number"),
+      value: z.number(),
+      formatOptions: z.object({...}).optional()
+    }))
+  )
+};
+```
+
+One interesting thing I discovered was how the AI learned to use the tools correctly. At first, it would make mistakes (passing flat arrays instead of 2D arrays, or using the wrong cell format). I added explicit error messages to help guide it, and the AI could actually self-correct at runtime. It would see the error, understand what went wrong, and retry with the correct format.
+
+But self-correction isn't the end goal. It's a signal that something needs fixing. Even though your AI assistant can recover from mistakes, you should be watching for recurring tool errors and using them as feedback to improve your tool definitions. I'd copy the incorrect tool call, the error message, and what the AI eventually sent that worked, then feed all that back to Claude and ask it to update the tool descriptions, parameter schemas, or validation logic to prevent that mistake from happening in the first place. So over time, the tools got better and better, and the AI needed to self-correct less and less. It's a development pattern I hadn't really used before: using AI to help you write better prompts and tool definitions for itself, turning runtime errors into design improvements.
+
+[Full code: src/tools/spreadsheet-tools.ts](https://github.com/michaelmagan/cheatsheet/blob/main/src/tools/spreadsheet-tools.ts)
+
+## Making Selections Part of the Context
+
+<video
+  src="https://github.com/user-attachments/assets/958dbd48-1b7d-4471-93a0-62125a685587"
+  controls
+/>
+
+For me, one of the most important UX patterns is letting users provide context to the AI through the GUI, not just text. I wanted to be able to select a range of cells and then say "sort this" or "sum this up" without having to type out "cells A1 through D5."
+
+Without selection context, you'd have to type something like: "Can you sort cells A1 through D5 by the values in column A?" That's clunky. With selection context, you just highlight the cells with your mouse and say "sort by column A." Much faster, much more natural. It's a great example of blending the best of graphical interfaces (direct manipulation) with natural language interfaces (flexibility and expressiveness).
+
+I used additional context helpers for this. It just passes the selected range with every message. Since the AI already has the full spreadsheet state from the markdown context helper, it can figure out exactly what I'm referring to.
```

Comment:
slightly awkward "... context *helpers* for this. *It* just passes.." - either singular or plural, just be consistent